### DISCUSSION TITLE
[Feature] Support `optimizePackageImports` like NextJS

### DISCUSSION THREAD

**Discussion by vimutti77** [remix-run/remix#7515] [Proposals]
## Problem
Some packages are export hundreds or thousands of modules (e.g., `@mui/material`, `@mui/icons-material`), which can cause performance issues in development and production.

This is the build time before I import from `@mui/icons-material`
![image](https://github.com/remix-run/remix/assets/27840664/2f9a5ed4-ea28-406d-9778-ef6f2932d22f)

This is the build time after I add `import { Delete } from "@mui/icons-material"`. It has increased about 20X.
![image](https://github.com/remix-run/remix/assets/27840664/5a9e2f5e-cf76-49f7-8be3-52adf25f3498)

This problem can be solved by changing the import from `top-level` import.
```ts
import { Delete } from '@mui/icons-material';
```
To `direct` import.
```ts
import Delete from '@mui/icons-material/Abc';
```
But this will have a huge impact on DX, as VSCode will always auto-suggest the `top-level` import. So, I have to edit the import path every time after I use the suggestion.
![image](https://github.com/remix-run/remix/assets/27840664/073c32e2-2cb9-4c7e-92c8-d140365f97de)

## Solution A
Automatically optimizes package imports similar to NextJS
https://nextjs.org/docs/app/api-reference/next-config-js/optimizePackageImports

## Solution B
Allow us to add the plugin to transform the code before the compilation. So, we can create and use some plugin like https://github.com/avocadowastaken/babel-plugin-direct-import

**Comment by jansedlon** [comment]
This might help? [vite-plugin-barrel](https://github.com/JiangWeixian/vite-plugin-barrel)
